### ISSUE TITLE
Inconsistency in Gradients Calculation of max_pooling and reduce_max

### ISSUE THREAD

**Issue by River861** [tensorflow/tensorflow#51682]
Hi, I found that the gradients calculations of `max_pooling` and `reduce_max`  are inconsistent in tensorflow, which make me really confused. The inconsistency exists when there are **multiple max elements**.

In `max_pooling`, only one of the max elements will get the gradients; while in `reduce_max`, all the max elements will divide the gradients equally. I want to know which algorithm is correct?

Here is an example code of the gradients calculation of `max_pooling` using tensorflow2.6.0:
```python
import numpy as np
import tensorflow as tf

with tf.GradientTape() as tape:
    x = tf.Variable([[[0.6],
                      [0.6],
                      [0.3]]])

    max_pool_1d = tf.keras.layers.MaxPooling1D(pool_size=3)
    y = max_pool_1d(x)

g = tape.gradient(y, x)


print("gradients of max: ", g.numpy())
```
The result is:
```
gradients of max:  [[[1.]
                     [0.]
                     [0.]]]
```
And this is the code of `reduce_max`:
```python
import numpy as np
import tensorflow as tf

with tf.GradientTape() as tape:
    x = tf.Variable([[[0.6],
                      [0.6],
                      [0.3]]])

    y = tf.reduce_max(x)
g = tape.gradient(y, x)


print("gradients of max: ", g.numpy())
```
The result is:
```
gradients of max:  [[[0.5]
                     [0.5]
                     [0. ]]]
```

Besides, I also found the gradients calculation of `max_pooling` in theano is different from either of the above methods. In theano's implementation, each of the max elements get the whole gradients, that is, the result of theano is:
```
gradients of max:  [[[1.]
                     [1.]
                     [0.]]]
```
Is it correct?

Any replies will be appreciated.
Thanks.

**Comment by Saduf2019** [comment]
@River861 
max_pooling is localized and and reduce_max is not  so they will not have same gradients.
You may open this issue at discussion forum as there is a larger community to respond there and as this is not a performance issue or bug.

**Comment by River861** [comment]
@Saduf2019,
I think I get that.  Thank you!

**Comment by Saduf2019** [comment]
@River861 
Can you please move this to close status as it answers your question.
